File: solutions/python/solution_03_bitset.py
ALPHABET = 26


def removeAnagramsAndSubAnagrams(words: list[str]) -> list[str]:
    if not words:
        return []

    groups: dict[tuple[int, ...], list[str]] = {}
    for word in words:
        key = tuple(freq_vector(word))
        groups.setdefault(key, []).append(word)

    unique = {k: v[0] for k, v in groups.items() if len(v) == 1}
    if not unique:
        return []

    vecs = sorted(unique.keys(), key=sum, reverse=True)
    max_freq = max(max(v) for v in vecs) + 1
    bitsets = [[0] * max_freq for _ in range(ALPHABET)]
    maximal = []

    for vec in vecs:
        if is_dominated(vec, bitsets):
            continue
        add_to_bitsets(vec, len(maximal), bitsets)
        maximal.append(vec)

    return [unique[vec] for vec in maximal]


def freq_vector(word: str) -> list[int]:
    freq = [0] * ALPHABET
    for c in word:
        freq[ord(c) - ord('a')] += 1
    return freq


def is_dominated(vec: tuple[int, ...], bitsets: list[list[int]]) -> bool:
    mask = ~0
    for letter in range(ALPHABET):
        need = vec[letter]
        if need == 0:
            continue
        mask &= bitsets[letter][need]
        if mask == 0:
            return False
    return mask != 0


def add_to_bitsets(vec: tuple[int, ...], idx: int, bitsets: list[list[int]]) -> None:
    bit = 1 << idx
    for letter in range(ALPHABET):
        for c in range(1, vec[letter] + 1):
            bitsets[letter][c] |= bit

File: solutions/kotlin/Solution03_Bitset.kt
private const val ALPHABET_03 = 26
private const val MAX_FREQ_03 = 17

fun removeAnagramsAndSubAnagrams03(words: List<String>): List<String> {
    if (words.isEmpty()) return emptyList()

    val groups = words.groupBy { freqVector03(it).toList() }
    val uniqueGroups = groups.filter { it.value.size == 1 }

    if (uniqueGroups.isEmpty()) return emptyList()

    val vecs = uniqueGroups.keys
        .map { it.toIntArray() }
        .sortedByDescending { it.sum() }

    val maximal = mutableListOf<IntArray>()
    val bitsets = Array(ALPHABET_03) { Array(MAX_FREQ_03) { mutableListOf<Long>() } }

    for (vec in vecs) {
        if (isDominated03(vec, maximal.size, bitsets)) continue
        addToBitsets03(vec, maximal.size, bitsets)
        maximal.add(vec)
    }

    return maximal.map { uniqueGroups[it.toList()]!!.first() }
}

private fun freqVector03(word: String): IntArray {
    val freq = IntArray(ALPHABET_03)
    for (c in word) {
        if (c in 'a'..'z') freq[c - 'a']++
    }
    return freq
}

private fun isDominated03(
    vec: IntArray,
    count: Int,
    bitsets: Array<Array<MutableList<Long>>>
): Boolean {
    if (count == 0) return false

    val blocks = (count + 63) / 64
    val mask = LongArray(blocks) { -1L }
    val lastBlockBits = count % 64
    if (lastBlockBits != 0) {
        mask[blocks - 1] = (1L shl lastBlockBits) - 1
    }

    for (letter in 0 until ALPHABET_03) {
        val need = vec[letter]
        if (need == 0) continue

        val bs = bitsets[letter][need]
        if (bs.isEmpty()) return false

        for (j in mask.indices) {
            mask[j] = mask[j] and (bs.getOrElse(j) { 0L })
        }

        if (mask.all { it == 0L }) return false
    }

    return true
}

private fun addToBitsets03(vec: IntArray, idx: Int, bitsets: Array<Array<MutableList<Long>>>) {
    val block = idx / 64
    val bit = 1L shl (idx % 64)

    for (letter in 0 until ALPHABET_03) {
        for (c in 1..vec[letter]) {
            val bs = bitsets[letter][c]
            while (bs.size <= block) bs.add(0L)
            bs[block] = bs[block] or bit
        }
    }
}
